utils/nfsd/nfsd.man
.\"
.\" nfsd(8)
.\"
.\" Copyright (C) 1999 Olaf Kirch <okir@monad.swb.de>
.TH rpc.nfsd 8 "20 Feb 2014"
.SH NAME
rpc.nfsd \- NFS server process
.SH SYNOPSIS
.BI "/usr/sbin/rpc.nfsd [" options "]" " "nproc
.SH DESCRIPTION
The
.B rpc.nfsd
program implements the user level part of the NFS service. The
main functionality is handled by the
.B nfsd
kernel module. The user space program merely specifies what sort of sockets
the kernel service should listen on, what NFS versions it should support, and
how many kernel threads it should use.
.P
The
.B rpc.mountd
server provides an ancillary service needed to satisfy mount requests
by NFS clients.
.SH OPTIONS
.TP
.B \-d " or " \-\-debug
enable logging of debugging messages
.TP
.B \-H " or " \-\-host  hostname
specify a particular hostname (or address) that NFS requests will
be accepted on. By default,
.B rpc.nfsd
will accept NFS requests on all known network addresses.
Note that
.B lockd
(which performs file locking services for NFS) may still accept
request on all known network addresses.  This may change in future
releases of the Linux Kernel. This option can be used multiple time 
to listen to more than one interface.
.TP
.B \-p " or " \-\-port  port
specify a different port to listen on for NFS requests. By default,
.B rpc.nfsd
will listen on port 2049.
.TP
.B \-r " or " \-\-rdma
specify that NFS requests on the standard RDMA port ("nfsrdma", port
20049) should be honored.
.TP
.BI \-\-rdma= port
Listen for RDMA requests on an alternate port - may be a number or a
name listed in
.BR /etc/services .
.TP
.B \-N " or " \-\-no-nfs-version vers
This option can be used to request that 
.B rpc.nfsd
does not offer certain versions of NFS. The current version of
.B rpc.nfsd
can support major NFS versions 2,3,4 and the minor versions 4.0, 4.1 and 4.2.
.TP
.B \-s " or " \-\-syslog
By default,
.B rpc.nfsd
logs error messages (and debug messages, if enabled) to stderr. This option makes 
.B rpc.nfsd
log these messages to syslog instead. Note that errors encountered during
option processing will still be logged to stderr regardless of this option.
.TP
.B \-T " or " \-\-no-tcp
Disable 
.B rpc.nfsd 
from accepting TCP connections from clients.
.TP
.B \-U " or " \-\-no-udp
Disable
.B rpc.nfsd
from accepting UDP connections from clients.
.TP
.B \-V " or " \-\-nfs-version vers
This option can be used to request that 
.B rpc.nfsd
offer certain versions of NFS. The current version of
.B rpc.nfsd
can support major NFS versions 2,3,4 and the minor versions 4.0, 4.1 and 4.2.
.TP
.B \-L " or " \-\-lease-time seconds
Set the lease-time used for NFSv4.  This corresponds to how often
clients need to confirm their state with the server. Valid range is
from 10 to 3600 seconds.
.TP
.B \-G " or " \-\-grace-time seconds
Set the grace-time used for NFSv4 and NLM (for NFSv2 and NFSv3).
New file open requests (NFSv4) and new file locks (NLM) will not be
allowed until after this time has passed to allow clients to recover state.
.TP
.I nproc
specify the number of NFS server threads. By default, eight
threads are started. However, for optimum performance several threads
should be used. The actual figure depends on the number of and the work
load created by the NFS clients, but a useful starting point is
eight threads. Effects of modifying that number can be checked using
the
.BR nfsstat (8)
program.
.P
Note that if the NFS server is already running, then the options for
specifying host, port, and protocol will be ignored.  The number of
processes given will be the only option considered, and the number of
active
.B nfsd
processes will be increased or decreased to match this number.
In particular
.B rpc.nfsd 0
will stop all threads and thus close any open connections.

.SH CONFIGURATION FILE
Many of the options that can be set on the command line can also be
controlled through values set in the
.B [nfsd]
section of the
.I /etc/nfs.conf
configuration file.  Values recognized include:
.TP
.B threads
The number of threads to start.
.TP
.B host
A host name, or comma separated list of host names, that
.I rpc.nfsd
will listen on.  Use of the
.B --host
option replaces all host names listed here.
.TP
.B grace-time
The grace time, for both NFSv4 and NLM, in seconds.
.TP
.B lease-time
The lease time for NFSv4, in seconds.
.TP
.B port
Set the port for TCP/UDP to bind to.
.TP
.B rdma
Set RDMA port.  Use "rdma=nfsrdma" to enable standard port.
.TP
.B UDP
Enable (with "on" or "yes" etc) or disable ("off", "no") UDP support.
.TP
.B TCP
Enable or disable TCP support.
.TP
.B vers2
.TP
.B vers3
.TP
.B vers4
Enable or disable a major NFS version.  3 and 4 are normally enabled
by default.
.TP
.B vers4.1
.TP
.B vers4.2
Setting these to "off" or similar will disable the selected minor
versions.  Setting to "on" will enable them.  The default values
are determined by the kernel, and usually minor versions default to
being enabled once the implementation is sufficiently complete.

.SH NOTES
If the program is built with TI-RPC support, it will enable any protocol and
address family combinations that are marked visible in the
.B netconfig
database.

.SH SEE ALSO
.BR nfsd (7),
.BR rpc.mountd (8),
.BR exports (5),
.BR exportfs (8),
.BR nfs.conf (5),
.BR rpc.rquotad (8),
.BR nfsstat (8),
.BR netconfig(5).
.SH AUTHOR
Olaf Kirch, Bill Hawes, H. J. Lu, G. Allan Morris III,
and a host of others.


NEWS
Significant changes for nfs-utils 1.1.0 - March/April 2007

 - rpc.lockd is gone.  One 3 old kernel releases need it.
 - /sbin/{u,}mount.nfs{,4} are now installed so 'mount' will
   use these to mount nfs filesystems instead of internal code.
  + mount.nfs will check for 'statd' to be running when mounting
    a filesystem which requires it.  If it is not running it will
    run "/usr/sbin/start-statd" to try to start it.
    If statd is not running and cannot be started, mount.nfs will
    refuse to mount the filesystem and will suggest the 'nolock'
    option.
 - Substantial changes to statd
  + The 'notify' process that must happen at boot has been split
    into a separate program "sm-notify".  It ensures that it
    only runs once even if you restart statd.  This is correct
    behaviour.
  + statd stores state in the files in /var/lib/nfs/sm/ so that
    if you kill and restart it, it will restore that state and
    continue working correctly.
  + statd makes more use of DNS lookup and should handle
    multi-homed peers better.  In particular, files in
    /var/lib/nfs/sm/ are named with the Full Qualified Domain Name
    if available.
 - If you export a directory as 'crossmnt', all filesystems
   mounted beneath are automatically exported with the same
   options (unless explicitly exported with different options).
 - subtree_check is no-longer the default.  The default is now
   no_subtree_check.
 - By default the system 'rpcgen' is used while building
   nfs-utils rather than the internal one.
 - Exportfs will warn if you try to export a filesystem that does
   not support NFS export.
 - Comprehensive notes on startup dependencies have been added
   to the README file.
 - Mount and statd now listen on a non-privileged port by default.
   For maximum safety an upgrade to portmap is recommended.
        git://neil.brown.name/portmap

 - This release should work with MIT Kerberos and Heimdal 0.8.1 and later.


 - A new option, -n, was added to rpc.gssd which specifies that
   accesses by root should not use 'machine credentials' when
   accessing NFS file systems mounted with Kerberos.  Using this
   option allows the root user to access the NFS space using any
   Kerberos principal, rather than always using the machine
   credentials.  However, its use also requires that root manually
   authenticate before attempting a mount with Kerberos.

   When rpc.gssd uses machine credentials, the selection algorithm has
   been changed.  Instead of simply using the first "nfs/*" key in the
   keytab, the keytab is now searched for keys in the following
   defined order:

     root/<fqdn>@REALM
     nfs/<fqdn>@REALM
     host/<fqdn>@REALM
     root/<any-name>@REALM
     nfs/<any-name>@REALM
     host/<any-name>@REALM


utils/nfsd/nfsd.c
/*
 * nfsd
 *
 * This is the user level part of nfsd. This is very primitive, because
 * all the work is now done in the kernel module.
 *
 * Copyright (C) 1995, 1996 Olaf Kirch <okir@monad.swb.de>
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>
#include <getopt.h>
#include <netdb.h>
#include <libgen.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include "conffile.h"
#include "nfslib.h"
#include "nfssvc.h"
#include "xlog.h"
#include "xcommon.h"

#ifndef NFSD_NPROC
#define NFSD_NPROC 8
#endif

char *conf_path = NFS_CONFFILE;

static void	usage(const char *);

static struct option longopts[] =
{
	{ "host", 1, 0, 'H' },
	{ "help", 0, 0, 'h' },
	{ "no-nfs-version", 1, 0, 'N' },
	{ "nfs-version", 1, 0, 'V' },
	{ "tcp", 0, 0, 't' },
	{ "no-tcp", 0, 0, 'T' },
	{ "udp", 0, 0, 'u' },
	{ "no-udp", 0, 0, 'U' },
	{ "port", 1, 0, 'P' },
	{ "port", 1, 0, 'p' },
	{ "debug", 0, 0, 'd' },
	{ "syslog", 0, 0, 's' },
	{ "rdma", 2, 0, 'R' },
	{ "grace-time", 1, 0, 'G'},
	{ "lease-time", 1, 0, 'L'},
	{ NULL, 0, 0, 0 }
};

int
main(int argc, char **argv)
{
	int	count = NFSD_NPROC, c, i, error = 0, portnum, fd, found_one;
	char *p, *progname, *port, *rdma_port = NULL;
	char **haddr = NULL;
	int hcounter = 0;
	struct conf_list *hosts;
	int	socket_up = 0;
	unsigned int minorvers = 0;
	unsigned int minorversset = 0;
	unsigned int minormask = 0;
	unsigned int versbits = NFSCTL_VERDEFAULT;
	unsigned int protobits = NFSCTL_PROTODEFAULT;
	int grace = -1;
	int lease = -1;

	progname = basename(argv[0]);
	haddr = xmalloc(sizeof(char *));
	haddr[0] = NULL;

	xlog_syslog(0);
	xlog_stderr(1);

	conf_init();
	xlog_from_conffile("nfsd");
	count = conf_get_num("nfsd", "threads", count);
	grace = conf_get_num("nfsd", "grace-time", grace);
	lease = conf_get_num("nfsd", "lease-time", lease);
	port = conf_get_str("nfsd", "port");
	if (!port)
		port = "nfs";
	rdma_port = conf_get_str("nfsd", "rdma");
	if (conf_get_bool("nfsd", "udp", NFSCTL_UDPISSET(protobits)))
		NFSCTL_UDPSET(protobits);
	else
		NFSCTL_UDPUNSET(protobits);
	if (conf_get_bool("nfsd", "tcp", NFSCTL_TCPISSET(protobits)))
		NFSCTL_TCPSET(protobits);
	else
		NFSCTL_TCPUNSET(protobits);
	for (i = 2; i <= 4; i++) {
		char tag[10];
		sprintf(tag, "vers%d", i);
		if (conf_get_bool("nfsd", tag, NFSCTL_VERISSET(versbits, i)))
			NFSCTL_VERSET(versbits, i);
		else
			NFSCTL_VERUNSET(versbits, i);
	}

	nfssvc_get_minormask(&minormask);
	/* We assume the kernel will default all minor versions to 'on',
	 * and allow the config file to disable some.
	 */
	if (NFSCTL_VERISSET(versbits, 4)) {
		NFSCTL_MINORSET(minorversset, 0);
		NFSCTL_MINORSET(minorvers, 0);
	}
	for (i = 1; i <= NFS4_MAXMINOR; i++) {
		char tag[20];
		sprintf(tag, "vers4.%d", i);
		/* The default for minor version support is to let the
		 * kernel decide.  We could ask the kernel what that choice
		 * will be, but that is needlessly complex.
		 * Instead, perform a config-file lookup using each of the
		 * two possible default.  If the result is different from the
		 * default, then impose that value, else don't make a change
		 * (i.e. don't set the bit in minorversset).
		 */
		if (!conf_get_bool("nfsd", tag, 1)) {
			NFSCTL_MINORSET(minorversset, i);
			NFSCTL_MINORUNSET(minorvers, i);
		}
		if (conf_get_bool("nfsd", tag, 0)) {
			NFSCTL_MINORSET(minorversset, i);
			NFSCTL_MINORSET(minorvers, i);
		}
	}

	hosts = conf_get_list("nfsd", "host");
	if (hosts && hosts->cnt) {
		struct conf_list_node *n;
		haddr = realloc(haddr, sizeof(char*) * hosts->cnt);
		TAILQ_FOREACH(n, &(hosts->fields), link) {
			haddr[hcounter] = n->field;
			hcounter++;
		}
	}

	while ((c = getopt_long(argc, argv, "dH:hN:V:p:P:stTitUrG:L:", longopts, NULL)) != EOF) {
		switch(c) {
		case 'd':
			xlog_config(D_ALL, 1);
			break;
		case 'H':
			if (hosts) {
				hosts = NULL;
				hcounter = 0;
			}
			if (hcounter) {
				haddr = realloc(haddr, sizeof(char*) * hcounter+1);
				if(!haddr) {
					fprintf(stderr, "%s: unable to allocate "
							"memory.\n", progname);
					exit(1);
				}
			}
			haddr[hcounter] = optarg;
			hcounter++;
			break;
		case 'P':	/* XXX for nfs-server compatibility */
		case 'p':
			/* only the last -p option has any effect */
			port = optarg;
			break;
		case 'r':
			rdma_port = "nfsrdma";
			break;
		case 'R': /* --rdma */
			if (optarg)
				rdma_port = optarg;
			else
				rdma_port = "nfsrdma";
			break;

		case 'N':
			switch((c = strtol(optarg, &p, 0))) {
			case 4:
				if (*p == '.') {
					int i = atoi(p+1);
					if (i < 0 || i > NFS4_MAXMINOR) {
						fprintf(stderr, "%s: unsupported minor version\n", optarg);
						exit(1);
					}
					NFSCTL_MINORSET(minorversset, i);
					NFSCTL_MINORUNSET(minorvers, i);
					if (minorvers != 0)
						break;
				} else {
					minorvers = 0;
					minorversset = minormask;
				}
			case 3:
			case 2:
				NFSCTL_VERUNSET(versbits, c);
				break;
			default:
				fprintf(stderr, "%s: Unsupported version\n", optarg);
				exit(1);
			}
			break;
		case 'V':
			switch((c = strtol(optarg, &p, 0))) {
			case 4:
				if (*p == '.') {
					int i = atoi(p+1);
					if (i < 0 || i > NFS4_MAXMINOR) {
						fprintf(stderr, "%s: unsupported minor version\n", optarg);
						exit(1);
					}
					NFSCTL_MINORSET(minorversset, i);
					NFSCTL_MINORSET(minorvers, i);
				} else
					minorvers = minorversset = minormask;
			case 3:
			case 2:
				NFSCTL_VERSET(versbits, c);
				break;
			default:
				fprintf(stderr, "%s: Unsupported version\n", optarg);
				exit(1);
			}
			break;
		case 's':
			xlog_syslog(1);
			xlog_stderr(0);
			break;
		case 't':
			NFSCTL_TCPSET(protobits);
			break;
		case 'T':
			NFSCTL_TCPUNSET(protobits);
			break;
		case 'u':
			NFSCTL_UDPSET(protobits);
			break;
		case 'U':
			NFSCTL_UDPUNSET(protobits);
			break;
		case 'G':
			grace = strtol(optarg, &p, 0);
			if (*p || grace <= 0) {
				fprintf(stderr, "%s: Unrecognized grace time.\n", optarg);
				exit(1);
			}
			break;
		case 'L':
			lease = strtol(optarg, &p, 0);
			if (*p || lease <= 0) {
				fprintf(stderr, "%s: Unrecognized lease time.\n", optarg);
				exit(1);
			}
			break;
		default:
			fprintf(stderr, "Invalid argument: '%c'\n", c);
		case 'h':
			usage(progname);
		}
	}

	if (optind < argc) {
		if ((count = atoi(argv[optind])) < 0) {
			/* insane # of servers */
			fprintf(stderr,
				"%s: invalid server count (%d), using 1\n",
				argv[0], count);
			count = 1;
		} else if (count == 0) {
			/*
			 * don't bother setting anything else if the threads
			 * are coming down anyway.
			 */
			socket_up = 1;
			goto set_threads;
		}
	}

	xlog_open(progname);

	portnum = strtol(port, &p, 0);
	if (!*p && (portnum <= 0 || portnum > 65535)) {
		/* getaddrinfo will catch other errors, but not
		 * out-of-range numbers.
		 */
		xlog(L_ERROR, "invalid port number: %s", port);
		exit(1);
	}

	/* make sure that at least one version is enabled */
	found_one = 0;
	for (c = NFSD_MINVERS; c <= NFSD_MAXVERS; c++) {
		if (NFSCTL_VERISSET(versbits, c))
			found_one = 1;
	}
	if (!found_one) {
		xlog(L_ERROR, "no version specified");
		exit(1);
	}

	if (NFSCTL_VERISSET(versbits, 4) &&
	    !NFSCTL_TCPISSET(protobits)) {
		xlog(L_ERROR, "version 4 requires the TCP protocol");
		exit(1);
	}

	if (chdir(NFS_STATEDIR)) {
		xlog(L_ERROR, "chdir(%s) failed: %m", NFS_STATEDIR);
		exit(1);
	}

	/* make sure nfsdfs is mounted if it's available */
	nfssvc_mount_nfsdfs(progname);

	/* can only change number of threads if nfsd is already up */
	if (nfssvc_inuse()) {
		socket_up = 1;
		goto set_threads;
	}

	/*
	 * Must set versions before the fd's so that the right versions get
	 * registered with rpcbind. Note that on older kernels w/o the right
	 * interfaces, these are a no-op.
	 * Timeouts must also be set before ports are created else we get
	 * EBUSY.
	 */
	nfssvc_setvers(versbits, minorvers, minorversset);
	if (grace > 0)
		nfssvc_set_time("grace", grace);
	if (lease  > 0)
		nfssvc_set_time("lease", lease);

	i = 0;
	do {
		error = nfssvc_set_sockets(protobits, haddr[i], port);
		if (!error)
			socket_up = 1;
	} while (++i < hcounter);

	if (rdma_port) {
		error = nfssvc_set_rdmaport(rdma_port);
		if (!error)
			socket_up = 1;
	}
set_threads:
	/* don't start any threads if unable to hand off any sockets */
	if (!socket_up) {
		xlog(L_ERROR, "unable to set any sockets for nfsd");
		goto out;
	}
	error = 0;

	/*
	 * KLUDGE ALERT:
	 * Some kernels let nfsd kernel threads inherit open files
	 * from the program that spawns them (i.e. us).  So close
	 * everything before spawning kernel threads.  --Chip
	 */
	fd = open("/dev/null", O_RDWR);
	if (fd == -1)
		xlog(L_ERROR, "Unable to open /dev/null: %m");
	else {
		/* switch xlog output to syslog since stderr is being closed */
		xlog_syslog(1);
		xlog_stderr(0);
		(void) dup2(fd, 0);
		(void) dup2(fd, 1);
		(void) dup2(fd, 2);
	}
	closeall(3);

	if ((error = nfssvc_threads(count)) < 0)
		xlog(L_ERROR, "error starting threads: errno %d (%m)", errno);
out:
	free(haddr);
	return (error != 0);
}

static void
usage(const char *prog)
{
	fprintf(stderr, "Usage:\n"
		"%s [-d|--debug] [-H hostname] [-p|-P|--port port]\n"
		"   [-N|--no-nfs-version version] [-V|--nfs-version version]\n"
		"   [-s|--syslog] [-t|--tcp] [-T|--no-tcp] [-u|--udp] [-U|--no-udp]\n"
		"   [-r|--rdma=] [-G|--grace-time secs] [-L|--leasetime secs] nrservs\n",
		prog);
	exit(2);
}
